CMakeLists.txt

```diff
@@ -1,504 +1,2030 @@
+    string (REGEX REPLACE "[-/]MD[d]*" " " ${var_} "${${var_}}")
+  endforeach ()
+
+  target_compile_options (common
+    INTERFACE
+      -bigobj            # Increase object file max size
+      -fp:precise        # Floating point behavior
+      -Gd                # __cdecl calling convention
+      -Gm-               # Minimal rebuild: disabled
+      -Gy-               # Function level linking: disabled
+      -MP                # Multiprocessor compilation
+      -openmp-           # pragma omp: disabled
+     #-FS                # forces PDB writes through server.-MP enables this by default
+      -errorReport:none  # No error reporting to Internet
+      -nologo            # Suppress login banner
+      -wd4018            # Disable signed/unsigned comparison warnings
+      -wd4244            # Disable float to int possible loss of data warnings
+      -wd4267            # Disable size_t to T possible loss of data warnings
+      -wd4800            # Disable C4800(int to bool performance)
+      -wd4503            # Decorated name length exceeded, name was truncated
+      $<$<COMPILE_LANGUAGE:CXX>:
+        -EHa
+        -GR
+      >
+      $<$<CONFIG:Release>:-Ox>
+      $<$<AND:$<COMPILE_LANGUAGE:CXX>,$<CONFIG:Debug>>:
+        -GS
+        -Zc:forScope
+      >
+      # static runtime
+      $<$<CONFIG:Debug>:-MTd>
+      $<$<NOT:$<CONFIG:Debug>>:-MT>
+      $<$<BOOL:${werr}>:-WX>
+      )
+  target_compile_definitions (common
+    INTERFACE
+      _WIN32_WINNT=0x6000
+      _SCL_SECURE_NO_WARNINGS
+      _CRT_SECURE_NO_WARNINGS
+      WIN32_CONSOLE
+      NOMINMAX
+      $<$<AND:$<COMPILE_LANGUAGE:CXX>,$<CONFIG:Debug>>:_CRTDBG_MAP_ALLOC>)
+  target_link_libraries (common
+    INTERFACE
+      -errorreport:none
+      -machine:X64)
+else ()
+  # HACK : because these need to come first, before any warning demotion
+  string (APPEND CMAKE_CXX_FLAGS "-Wall -Wdeprecated")
+  # not MSVC
+  target_compile_options (common
+    INTERFACE
+      $<$<BOOL:${werr}>:-Werror>
+      $<$<COMPILE_LANGUAGE:CXX>:
+        -frtti
+        -Wno-invalid-offsetof
+        -Wnon-virtual-dtor
+      >
+      -Wno-sign-compare
+      -Wno-char-subscripts
+      -Wno-format
+      -Wno-unused-local-typedefs
+      $<$<BOOL:${is_gcc}>:
+        -Wno-unused-but-set-variable
+        -Wno-deprecated
+      >
+      $<$<BOOL:${is_clang}>:
+        -Wno-mismatched-tags
+        -Wno-deprecated-register
+      >
+      $<$<AND:$<BOOL:${is_clang}>,$<COMPILE_LANGUAGE:CXX>>:
+        -Wno-redeclared-class-member
+      >
+      $<$<NOT:$<CONFIG:Debug>>:-fno-strict-aliasing>
+      # tweak gcc optimization for debug
+      $<$<AND:$<BOOL:is_gcc>,$<CONFIG:Debug>>:-O0>
+      # Add debug symbols to release config
+      $<$<CONFIG:Debug>:-g>)
+  target_compile_definitions (common
+    INTERFACE
+      _FILE_OFFSET_BITS=64)
+  target_link_libraries (common INTERFACE -rdynamic)
+endif ()
+
+if (static)
+  target_link_libraries (common INTERFACE -static-libstdc++)
```

Comment:
This and the other things messing with `CMAKE_CXX_FLAGS` are good contenders for a toolchain file. But I do think it can be easily phased as the next step as this is already a big PR as is. Maybe just put it as a GH issue or are you already tracking these things internally in JIRA?

Reply:
I agree about the toolchain (see comment line 188). we track issues internally right now, but this might change in the future